cdap-security/src/main/java/io/cdap/cdap/security/auth/TinkCipher.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.security.auth;
+
+import com.google.crypto.tink.Aead;
+import com.google.crypto.tink.CleartextKeysetHandle;
+import com.google.crypto.tink.JsonKeysetReader;
+import com.google.crypto.tink.KeysetHandle;
+import com.google.crypto.tink.aead.AeadConfig;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.conf.SConfiguration;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.Base64;
+
+/**
+ * Tink cipher that allows encrypting and decrypting data using keyset stored in {@link SConfiguration}
+ */
+public class TinkCipher {
+  /**
+   * Wraps a keyset with some additional parameters and metadata.
+   */
+  private final KeysetHandle keysetHandle;
+
+  /**
+   * Tink AEAD (Authenticated Encryption with Associated Data) primitive
+   */
+  private final Aead aead;
+
+  public TinkCipher(SConfiguration sConf) throws CipherException {
+    try {
+      // Init Tink with AEAD primitive
+      AeadConfig.register();
+
+      // Load keyset from sConf
+      String jsonKeyset = sConf.get(Constants.Security.Authentication.USER_CREDENTIAL_ENCRYPTION_KEYSET);
+      this.keysetHandle = CleartextKeysetHandle.read(JsonKeysetReader.withString(jsonKeyset));
+
+      this.aead = keysetHandle.getPrimitive(Aead.class);
+    } catch (GeneralSecurityException e) {
+      throw new CipherException("Failed to init Tink cipher", e);
+    } catch (IOException e) {
+      throw new CipherException("Failed to load Tink keyset", e);
+    }
+  }
+
+  /**
+   * Encrypt the data.
+   *
+   * @param plainData data to be encrypted
+   * @param associatedData used for integrity checking during decryption.
+   * @return encrypted data
+   * @throws CipherException if encryption fails
+   */
+  public byte[] encrypt(byte[] plainData, byte[] associatedData) throws CipherException {
```

Comment:
I understand that we're providing a direct interface adapter for the Tink library here, but do we really foresee a case in which we actually need to use the associatedData field?

If not, we may want to consider removing it from the interface to prevent accidental usage since associatedData is authenticated but not encrypted.

Reply:
Plan to use user id as associated data. Not absolutely needed. 

Make the associated data optional (e.g. nullable)

Reply:
Ack sounds good, if that's the case we need to make sure we're passing the associated data directly from the principal in the DefaultAccessEnforcer.